src/async-components/views/dialogs/security/ExportE2eKeysDialog.tsx

```diff
@@ -63,21 +68,40 @@ export default class ExportE2eKeysDialog extends React.Component<IProps, IState>
         this.unmounted = true;
     }
 
-    private onPassphraseFormSubmit = (ev: React.FormEvent): boolean => {
-        ev.preventDefault();
+    private async verifyFieldsBeforeSubmit(): Promise<boolean> {
+        const fieldIdsInDisplayOrder = [this.fieldPassword, this.fieldPasswordConfirm];
 
-        const passphrase = this.state.passphrase1;
-        if (passphrase !== this.state.passphrase2) {
-            this.setState({ errStr: _t("Passphrases must match") });
-            return false;
+        const invalidFields: Field[] = [];
+
+        for (const field of fieldIdsInDisplayOrder) {
+            if (!field) continue;
+
+            const valid = await field.validate({ allowEmpty: false });
+            if (!valid) {
+                invalidFields.push(field);
+            }
         }
-        if (!passphrase) {
-            this.setState({ errStr: _t("Passphrase must not be empty") });
-            return false;
+
+        if (invalidFields.length === 0) {
+            return true;
         }
 
-        this.startExport(passphrase);
+        // Focus on the first invalid field, then re-validate,
+        // which will result in the error tooltip being displayed for that field.
+        invalidFields[0].focus();
+        invalidFields[0].validate({ allowEmpty: false, focused: true });
+
         return false;
+    }
+
+    private onPassphraseFormSubmit = async (ev: React.FormEvent): Promise<void> => {
+        ev.preventDefault();
+
+        if (!(await this.verifyFieldsBeforeSubmit())) return;
+
+        const passphrase = this.state.passphrase1;
```

Comment:
isn't this racy? what happens if `passphrase1`  changes while passphrase2 is being validated?

(Or, for that matter, if the component is unmounted?)

Reply:
Unmounted is a valid point, re validation race that applies to all of our validated Fields. This pattern is used all over the shop, specifically these fields are also used during registration & password reset with the same supporting code

Reply:
I'm not sure I find that reassuring (or a particularly compelling reason not to do better here: eg have `verifyFieldsBeforeSubmit` return the verified state), but ok.